### PR TITLE
Particle Ancestry Map

### PR DESCRIPTION
**Pull request by henrylay97**
Currently in the largeant processes (as of ~1 year ago)  we are producing a `std::map<int, std::set<int>>` which indicates the mapping of dropped track IDs (the set of ints) to their undropped ancestor (the key). This object should allow you to find the intended ancestor if you have a track ID for which you cannot access the `MCParticle`.

This object has a few flaws:
- It doesn't have a name, which makes it very hard for analysers to know what it does (unless you know in advance).
- It cannot be _kept_ or _dropped_ using the output and input commands in fcl (as a result of not having a name that can be recognised by these prescriptions).
- When multiple instances of largeant are ran, the [MergeSimSources (larsim) module](https://github.com/LArSoft/larsim/blob/develop/larsim/MergeSimSources/MergeSimSources_module.cc) doesn't merge these objects so you lose it from the event. Running multiple largeant instances is a pretty common approach in SBND (and I'm sure in other experiments).
- Even if you hack the fcl file to keep the original maps, the track ID offsets applied in `MergeSimSources` are lost, so the maps values are garbage.

This PR produces a wrapper object in order to solve all of the above flaws. This also provides a few useful functions to access the information.

It will be accompanied by PRs in larg4 and larsim in order to produce the object, and have the ability to merge it.

In SBND this object isn't currently used in develop (I ran across this issue in a feature branch, which will later be prepared for develop) so shouldn't be a breaking change. The CI should help indicate whether any other repos use this object and will therefore need to update their usage.

### CODE REVIEW

**Comment by knoepfel** [comment]
Hi @henrylay97, thanks for the PR.  Before we review the code:

- When you say the `std::map<int, std::set<int>>` doesn't have a name, what do you mean?
- For the "keep" and "drop" commands, every C++ type has a "friendly" name, which translates the above type into something that doesn't contain angle brackets (`<`, `>`), commas, or underscores.  Unfortunately that "friendly" name isn't always so friendly.  

The pattern for dropping products using "keep" and "drop" looks like:
```
<friendly C++ name>_<module label>_<instance name>_<process name>
```
and it's possible to use wildcards to simplify the specification (e.g.):
```
drop *std::map*_larG4main_*_*
```
The above command is interpreted by the framework to drop all products with a friendly name that contains `std::map` as long as those products were created by the module with the label `larG4main`.  I'm guessing this is sufficient for getting you what you need in terms of "naming".  Of course, if you need the type to do more than just what a `std::map` does, then it may make sense to create a separate class as you have done.

**Comment by henrylay97** [comment]
Hi Kyle!

I wasn't aware of this C++ friendly name, that's useful to know, thank you! 

While you're right that it does provide the needed "naming" for these procedures, I think the readability point is still pretty important. 

Every other object on our events has a name that tells you, to some degree, what it does / represents or at least means you can look at a header file where a slightly lengthier explanation exists. This avoids analysers having to understand entire producer modules to decode the use of an object. It also allows us to provide functions that also quickly indicate their purpose.

I'm open to suggestions / other approaches.

File: lardataobj/Simulation/ParticleAncestryMap.cxx [hunk 1]
```diff
@@ -0,0 +1,41 @@
+#ifndef PARTICLEANCESTRYMAP_CXX
+#define PARTICLEANCESTRYMAP_CXX
+
+#include "lardataobj/Simulation/ParticleAncestryMap.h"
+
+namespace sim {
+
+  void ParticleAncestryMap::SetMap(const std::map<int, std::set<int>>& map) { fParticleMap = map; }
+
+  std::map<int, std::set<int>> ParticleAncestryMap::GetMap() const { return fParticleMap; }
```
Comment (knoepfel) [inline lardataobj/Simulation/ParticleAncestryMap.cxx]: Recommend returning by `const` reference to avoid unnecessary copies.

```suggestion
  std::map<int, std::set<int>> const& ParticleAncestryMap::GetMap() const { return fParticleMap; }
```

File: lardataobj/Simulation/ParticleAncestryMap.cxx [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+  bool ParticleAncestryMap::HasDroppedDescendants(const int& trackid)
```
Comment (knoepfel) [inline lardataobj/Simulation/ParticleAncestryMap.cxx]: Here and elsewhere, `int` objects (which are smaller than pointers) should be passed in by value:

```suggestion
  bool ParticleAncestryMap::HasDroppedDescendants(const int trackid)
```

File: lardataobj/Simulation/ParticleAncestryMap.cxx [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+  {
+    return fParticleMap.count(trackid) != 0;
+  }
+
+  std::set<int> ParticleAncestryMap::GetAllDroppedDescendants(const int& trackid)
```
Comment (knoepfel) [inline lardataobj/Simulation/ParticleAncestryMap.cxx]:
```suggestion
  std::set<int> const& ParticleAncestryMap::GetAllDroppedDescendants(const int trackid)
```

File: lardataobj/Simulation/ParticleAncestryMap.cxx [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+    return false;
+  }
+
+  int ParticleAncestryMap::GetAncestor(const int& trackid)
+  {
+    for (auto const& [ancestor, descendants] : fParticleMap) {
+      if (descendants.count(trackid) != 0) return ancestor;
+    }
+
+    return -std::numeric_limits<int>::max();
```
Comment (knoepfel) [inline lardataobj/Simulation/ParticleAncestryMap.cxx]: How do you expect users to use the value of `GetAncestor()`?  I can think of a few options:
```C++
// Option 1
if (int i = map->GetAncestor(42); i == - std::numeric_limits<int>::max()) { 
  // Ancestor doesn't exist.  Fragile due to comparison of hard-coded number
}

// Option 2
if (int i = map->GetAncestor(42); i < 0) {
  // Ancestor doesn't exist.  
  // Better due to not having to compare against hard-coded number.
}

// Option 3
if (int i = map->GetAncestor(42); !map->Exists(i)) {
  // Ancestor doesn't exist.  
  // Best option as user doesn't have to compare against any number but lets the implementation
  // decide how to do the comparison.
}
```

I suggestion using option 3, but that requires adding a new function.

File: lardataobj/Simulation/classes_def.xml [hunk 5]
```diff
@@ -104,8 +104,6 @@
  <class name="std::vector<sim::OpDetBacktrackerRecord>"/>
  <class name="std::vector<sim::SupernovaTruth>"/>
  <class name="std::vector<sim::SimDriftedElectronCluster>"/>
- <class name="std::pair<int,std::set<int>>"/>
```
Comment (knoepfel) [inline lardataobj/Simulation/classes_def.xml:107]: Why did you remove the `std::pair<int, std::set<int>>` dictionary?

**Review by knoepfel — COMMENTED** [review]
@henrylay97, your comment regarding naming is a fair one.  Your approach thus seems reasonable.  Please consider making the changes suggested below.  Thanks for the PR.

File: lardataobj/Simulation/ParticleAncestryMap.cxx (hunk 4 above)
Comment (henrylay97) [reply lardataobj/Simulation/ParticleAncestryMap.cxx]: Is this not covered by the `bool HasAncestor(const int trackid)` function?

File: lardataobj/Simulation/classes_def.xml (hunk 5 above)
Comment (henrylay97) [reply lardataobj/Simulation/classes_def.xml:107]: No idea, great question! I think that's just sloppy committing, sorry!

File: lardataobj/Simulation/ParticleAncestryMap.cxx (hunk 4 above)
Comment (knoepfel) [reply lardataobj/Simulation/ParticleAncestryMap.cxx]: Sorry for the delay in replying, @henrylay97.  Yes, the `HasAncestor(int trackid)` function already can return `true` or `false`.  The difference here, though, is that the code has already traversed the `std::map` when calling `map->GetAncestor(42)` and already knows that there is not an entry for 42.  The implementation of `Exists(int trackid)` would simply be something like:

```C++
bool Exists(int const trackid) const { 
  return trackid == -std::numeric_limits<int>::max(); 
}
```

This way, if you want to call `GetAncestor`, you don't have to traverse the tree again with `HasAncestor`.

Comment (henrylay97) [reply lardataobj/Simulation/ParticleAncestryMap.cxx]: I see! I misunderstood your point I think. I will implement an `Exists` function and probably remove the existing `HasAncestor` function to avoid confusion.

**Comment by lgarren** [comment]
trigger build with pull requests #34, LArSoft/larg4#45, LArSoft/larsim#110

**Comment by lgarren** [comment]
approve
